### DISCUSSION TITLE
Handling NULL values

### DISCUSSION THREAD

**Discussion by atqnp** [EvgSkv/logica#118] [Q&A] [answered]
Hi,

I'm sorry if this has been asked before.
How do I handle ```NULL``` values? For example:
```SQL
SELECT * FROM db WHERE a IS NOT NULL
```
or
```SQL
SELECT * FROM db WHERE a IS NULL
```
is it by just writing blanks when extracting data?
ex: ```role ==  ```

**Comment by jedi1992** [comment, marked as answer]
I dont know how to  create table(predicate) in logica with null values for the columns. 
Tried different things like None, Null, NULL, NULL() etc but could not. 

Assuming you are able to do it,  then using IsNull(role)  == true; works.  

 p.s IsNull is the camel case of ISNULL big query function

**Reply by atqnp** [reply]
Thank you very much for the reply!

I tried your suggestion but it needs a small tweak.
I am able to create a predicate using NULL. You have to make sure they are lowercases. I noticed it when I made a mistake when trying to handle a boolean value. Example: `True` will not work but `true` will.

Example of predicate with the answer (This is kind of a similar example in CoLab ):
```python
%%logica NoRole

@Engine("sqlite");

Employee(name: "Alice", role: "Product Manager");
Employee(name: "Bob", role: "Engineer");
Employee(name: "Caroline", role: "Engineer");
Employee(name: "David", role: "Data Scientist");
Employee(name: "Eve", role: null);

NoRole(name:) :- Employee(name:, role:),
  IsNull(role);
```

for NOT NULL, I had to put a negation `~IsNull(role);` This works but I don't know if there is a better way.

**Reply by EvgSkv** [reply]
@jedi1992 Thanks for answering this! @atqnp Great that you figured it out!

Just to confirm:
1) Boolean and null literals are indeed lowercase: `true`, `false`, `null`.
2) Checking for null is done with `IsNull(x)` function which compiles to SQL's `x IS NULL`.
3) Indeed for `is not null` we need to negate `IsNull`. It's preferred to negate it with `!` operator rather than `~`, however both works. I've [extended](https://colab.research.google.com/github/EvgSkv/logica/blob/main/tutorial/Logica_tutorial.ipynb#scrollTo=nvDm194jrFyc) section in the tutorial about it.
